Finalize egress task (#3222)
istio/istio.io: content/docs/tasks/traffic-management/egress/index.md

    {"level":"info","time":"2019-01-24T12:17:11.855496Z","instance":"accesslog.logentry.istio-system","apiClaims":"","apiKey":"","clientTraceId":"","connection_security_policy":"unknown","destinationApp":"","destinationIp":"I60GXg==","destinationName":"unknown","destinationNamespace":"default","destinationOwner":"unknown","destinationPrincipal":"","destinationServiceHost":"httpbin.org","destinationWorkload":"unknown","grpcMessage":"","grpcStatus":"","httpAuthority":"httpbin.org","latency":"214.661667ms","method":"GET","permissiveResponseCode":"none","permissiveResponsePolicyID":"none","protocol":"http","receivedBytes":270,"referer":"","reporter":"source","requestId":"17fde8f7-fa62-9b39-8999-302324e6def2","requestSize":0,"requestedServerName":"","responseCode":200,"responseSize":599,"responseTimestamp":"2019-01-24T12:17:11.855521Z","sentBytes":806,"sourceApp":"sleep","sourceIp":"AAAAAAAAAAAAAP//rB5tUg==","sourceName":"sleep-88ddbcfdd-rgk77","sourceNamespace":"default","sourceOwner":"kubernetes://apis/apps/v1/namespaces/default/deployments/sleep","sourcePrincipal":"","sourceWorkload":"sleep","url":"/headers","userAgent":"curl/7.60.0","xForwardedFor":"0.0.0.0"}
    {{< /text >}}

>     Note that the `destinationServiceHost` attribute is equal to `httpbin.org`. Also notice the HTTP-related attributes:

Author: @rcaballeromx the attributes appear on the following line: `method`, `url`, `responseCode`

}
{{< /text >}}

Note that this time you do not see any headers related to the Istio sidecar. Also note that the requests sent to

> external services appear neither in the log of the sidecar nor in the Mixer log: by bypassing Istio sidecars you lost

Author: @rcaballeromx I tried to be a bit dramatic here by using the term "you lose", thinking it is more stressful and is interesting to read. I see that you prefer neutral "you can not longer...".

monitoring of access to external services.

### Cleanup of direct access to external services

Update the `istio-sidecar-injector` configmap to redirect all outbound traffic to the sidecar proxies:

{{< text bash >}}
$ helm template install/kubernetes/helm/istio <the flags you used to install Istio> -x templates/sidecar-injector-configmap.yaml | kubectl apply -f -
{{< /text >}}

## Install Istio with access to all external services by default

An alternative to calling external services directly is to instruct the Istio proxy to pass through the calls to all the
external services. This option allows you to start evaluating Istio quickly, without controlling access to external
services, and decide to [configure access to external services](#controlled-access-to-external-services) later.

> Istio has an installation option that allows access to any external service on any ports that have no HTTP service and

Author: @rcaballeromx I added a link to Istio installation options - https://preliminary.istio.io/docs/reference/config/installation-options/

no service entry defined inside the mesh. For example, if there is no registered HTTP service or service entry defined for
port 8000 inside the mesh, the sidecar proxy can be configured to pass the request on to any external service on that

> port.

Author: Fixed in https://github.com/istio/istio.io/pull/3222/commits/821864b52b19e47e1c2b52ac5eaa04d4a8ab6021 .

In addition to that, you loose monitoring of access to external services and cannot apply Istio features on traffic to
external services.

The third approach directs the traffic through the Istio sidecar proxy, however it allows access to any service on any
ports that have no HTTP service inside the mesh and no service entry defined. In this approach, similarly to the

> previous one, you cannot monitor access to external services. The advantage of this approach is that you must not

Reviewer: is that you don't need to know which ...

Author: Fixed in https://github.com/istio/istio.io/pull/3222/commits/f2b596a392a6193f44132f9d7c8e9915ca63b33a

> know which IP ranges are external to the cluster. Additional advantage is that you can decide to switch to the first

Reviewer: Another advantage is that you can easily switch to the first approach for a specific port by simply creating a service entry for that port.

Author: Fixed in https://github.com/istio/istio.io/pull/3222/commits/c2ea3e3a4158a5f9d67868f42c69237b1c08f9f1.

approach per specific port, by creating service entries for that port. This way you can start Istio by allowing access

> to any external service and then decide to start controlling access to external services per specific port, gaining

Reviewer: for specific ports, enabling monitoring and ...

Author: Fixed in https://github.com/istio/istio.io/pull/3222/commits/df7ecc5dc81d43193267ed33c7d43f32c099d950

> monitoring and traffic control features later, once they are needed.

Reviewer: Maybe should also mention disadvantage/limitation is that this feature can't be used for some ports, like port 80?

Author: Addressed in https://github.com/istio/istio.io/pull/3222/commits/30bda5f61e845a50aae2ff9c864e4e67e465e59e, please take a look.

In addition to that, you loose monitoring of access to external services and cannot apply Istio features on traffic to
external services.

> The third approach directs the traffic through the Istio sidecar proxy, however it allows access to any service on any

Reviewer 2: ```suggestion
The third approach directs the traffic through the Istio sidecar proxy but it allows access to any service on any
```

> ports that have no HTTP service inside the mesh and no service entry defined. With this approach, like with the

Reviewer 2: ```suggestion
ports without an HTTP service in the mesh or without a service entry. Similarly to the second approach,
```

> previous one, you cannot monitor access to external services. The advantage of this approach is that you don't need to

Reviewer 2: ```suggestion
you cannot monitor the access to external services but you don't need to
```

> previous one, you cannot monitor access to external services. The advantage of this approach is that you don't need to

Reviewer: With this approach, like the previous one, you ...

Author: Addressed in https://github.com/istio/istio.io/pull/3222/commits/be148adab523d7a6b74db00580c28d71b255ec46: With this approach, like **with** the previous one

Reviewer 2: There are two previous approaches, I left a suggestion to be more specific and avoid confusion.

> know which IP ranges are external to the cluster. Another advantage is that you can easily switch to the first approach

Reviewer 2: ```suggestion
know which IP ranges are external to the cluster. Additionally, you can easily switch to the first approach
```

> for a specific port by simply creating a service entry for that port. This way you can start Istio by allowing access

Reviewer 2: ```suggestion
for a specific port: simply create a service entry for that port. You can use this approach to start using Istio allowing access
```

> to any external service and then decide to start controlling access to external services for specific ports, enabling

Reviewer 2: ```suggestion
to any external service and then decide to start controlling access to external services for specific ports.
```

> monitoring and traffic control features later, once they are needed. A caveat is that some ports, for example port 80,

Reviewer 2: ```suggestion
Then, you can enable traffic monitoring and control features once they are needed.  Some ports, for example port 80,
```

> already have HTTP services inside Istio by default, so for them only the first and the second approaches can be used.

Reviewer 2: ```suggestion
already have HTTP services inside Istio by default. Because of this caveat, you can only use the first and second approaches for those ports.
```

> already have HTTP services inside Istio by default, so for them only the first and the second approaches can be applied.

Reviewer: can be used.

Author: Fixed in https://github.com/istio/istio.io/pull/3222/commits/f22576017d824f20de8a9cb96cbcb08120edf42c
